### PR TITLE
fix(store): remove SelectType<T> implementation due to cannot assign @Select to private/protected property

### PR DESCRIPTION
**Pull request by splincode**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #1532 

## Does this PR introduce a breaking change?

```
[ ] Yes
[ ] No
```

### CODE REVIEW

**Comment by arturovt** [comment]
It fails because of `keyof` operator which tries to yield public property names in that construct that was added in this PR. AFAIK `keyof` doesn't yield private/protected property names.

But Max is more knowledgeable in that typing stuff so decisions are up to you.

**Comment by splincode** [comment]
@arturovt @markwhitfeld This is a big problem)) I have never used private/protected selectors. And we never had unit tests with private fields, so I thought that was normal behavior and therefore we introduced this in v3.6. While I do not really know how to fix this without breaking the behavior of the current version 3.6. What do you think can be fixed with this?

```ts
It fails because of keyof operator which tries to yield public property names 
```

Do you know how to pull out private/protected fields?

https://github.com/ngxs/store/blob/master/packages/store/src/decorators/select/symbols.ts#L48

**Comment by arturovt** [comment]
As TypeScript team says: **no way** :smile:

**Comment by splincode** [comment]
There is an offer to leave the `select` only for public fields)

**Comment by arturovt** [comment]
Btw what is the purpose of that type and can't we leave it as any?

P.S. just curious.

**Comment by splincode** [comment]
```
Btw what is the purpose of that type
```

- Auto inferable type from selector function

Example:

```ts
@State<Model>({
  name: 'my',
  default: {
     path: 'hello'
  }
})
class MyState {
  @Selector()
  public static getState(state) {
    return state.path;
  }
}
```

// MyState.getState() -> auto inferable (state: Model) => string

```ts
class MyComponent {
   @Select(MyState.getState) public path;  // Magic: auto inferable this.path: Observable<string>
}
```

**Comment by splincode** [comment]
```
can't we leave it as any
```

If move to any (v3.5)

```
class MyComponent {
   @Select(MyState.getState) public path;  // this.path: any
}
```

**Comment by arturovt** [comment]
I understand your point. But this seems like it has been a breaking change for users that might have private properties. Am I wrong?

**Comment by splincode** [comment]
```
But this seems like it has been a breaking change
```

Yes, but we have 3.6.0 for more than six months as a version in production)

**Comment by arturovt** [comment]
There is at least 1 guy that faced breaking changes :rofl:

**Comment by splincode** [comment]
image
we can add to the description that this was a critical change

**Comment by arturovt** [comment]
I'm actually not pushing things with this. As I said all decisions are up to you.

**Comment by splincode** [comment]
@markwhitfeld @arturovt I just don’t know how to fix this problem without breaking typing back (any)

**Comment by arturovt** [comment]
> There is at least 1 guy that faced breaking changes

image

Or even more.. At least there can be a quick fix, I mean just remove `private` and it'll be working.

**Comment by splincode** [comment]
```
I mean just remove private and it'll be working.
```

everyone who encountered this in my company did just that

**Comment by arturovt** [comment]
There is also a guy that says that he's having large codebase and it blocks him from upgrading.

Cannot advise anything in that case... I'm not a type expert :(

**Comment by splincode** [comment]
@markwhitfeld what we do? downgrade to any type?

**Comment by markwhitfeld** [comment]
Are we able to provide a fallback that will permit the private fields but then if the field is public then it will infer the type?
This is similar to https://github.com/DefinitelyTyped/DefinitelyTyped/issues/14811

**Comment by splincode** [comment]
```
Are we able to provide a fallback that will permit the private fields but then if the field is public then it will infer the type?
```

@markwhitfeld  tried, I didn’t succeed

**Comment by arturovt** [comment]
@splincode what do you think if we leave it as any? Personally, I think that your feature was great and I clearly understand its purpose. But we have to think about our users... :disappointed: 
Maybe we could add that feature for v4?

**Comment by markwhitfeld** [comment]
I think that we need to firstly address the regression, and if that means that we move back to `any` for now, then so be it!
We could investigate how to make this type safety work again and push it in another PR if we find a way.

**Comment by markwhitfeld** [comment]
PS. I thought that this construct only gave you the ability to detect if the types were mismatched (type safety). As far as I know, with TypeScript, a decorator cannot be used to infer a type of a property... unless this changed in a recent version.

**Comment by splincode** [comment]
@markwhitfeld sorry, you are right

**Comment by splincode** [comment]
@arturovt @markwhitfeld review again please

**Comment by kuncevic** [comment]
Hey guys, how about to consider of using schematic to fix up edge cases (breaking changes) like that automatically?

**Comment by markwhitfeld** [comment]
@kuncevic Nice idea, but I think that it is totally valid for someone to have a private observable. I don't think it would be right to adjust people's code to make it no longer private. Unless you were thinking about another approach?

**Comment by bbarry** [comment]
> 
> 
> @kuncevic Nice idea, but I think that it is totally valid for someone to have a private observable. I don't think it would be right to adjust people's code to make it no longer private. Unless you were thinking about another approach?

We declare all our store observables in the constructor in order to fully apply tsc strict compiler options:

```typescript
class MyComponent {
   private path: Observable<string>;
   constructor(store: Store) {
      this.path = store.select(MyState.getState);
   }
}
```

Perhaps that migration is a migration that could be done without changing the member accessibility.

**Comment by splincode** [comment]
@markwhitfeld updated

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,7 @@ $ npm install @ngxs/store@dev
 - Feature: Logger Plugin - Add filter for Logger Plugin [#1571](https://github.com/ngxs/store/pull/1571)
 - Performance: Logger Plugin - Plugin should lazy inject the store once [#1550](https://github.com/ngxs/store/pull/1550)
 - Fix: `ofAction*` methods should prevent passing anything except of `ActionType` [#1616](https://github.com/ngxs/store/pull/1616)
+- Fix: remove implement SelectType<T> due to cannot assign @Select to private/protected property [#1623](https://github.com/ngxs/store/pull/1623)
```
Comment (arturovt) [inline CHANGELOG.md:13]:
```suggestion
- Fix: remove SelectType<T> implementation due to cannot assign @Select to private/protected property [#1623](https://github.com/ngxs/store/pull/1623)
```

Comment (splincode) [reply CHANGELOG.md]: done

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,7 @@ $ npm install @ngxs/store@dev
 - Feature: Logger Plugin - Add filter for Logger Plugin [#1571](https://github.com/ngxs/store/pull/1571)
 - Performance: Logger Plugin - Plugin should lazy inject the store once [#1550](https://github.com/ngxs/store/pull/1550)
 - Fix: `ofAction*` methods should prevent passing anything except of `ActionType` [#1616](https://github.com/ngxs/store/pull/1616)
+- Fix: remove SelectType<T> implementation due to cannot assign @Select to private/protected property [#1623](https://github.com/ngxs/store/pull/1623)
```
Comment (markwhitfeld) [inline CHANGELOG.md:13]:
```suggestion
- Fix: Remove the recent `@Select` type safety check due to issues with private/protected properties [#1623](https://github.com/ngxs/store/pull/1623)
```

File: packages/store/types/tests/selection.lint.ts [hunk 3]
```diff
@@ -138,4 +144,43 @@ describe('[TEST]: Action Types', () => {
       }
     }
   });
+
+  it('issues/1532', () => {
```
Comment (markwhitfeld) [inline packages/store/types/tests/selection.lint.ts:148]:
```suggestion
  it('should support protected and private methods (https://github.com/ngxs/store/issues/1532)', () => {
```

**Comment by dmrickey** [comment]
I guess I've never tried assigning a select statement to a private property, but I'm just now jumping on the 3.7.x bandwagon from 3.6.2 and this is a heartbreaking change. I can see why someone might want to, but I would much rather have the typecheck back than be able to assign to private props this way.

**Comment by markwhitfeld** [comment]
@dmrickey I will be working on a better, more type-safe alternative to `@Select` very soon, so all is not lost here 👍 
This is the basic idea for syntax:
```TS
isLoggedIn$ = select(AuthState.isLoggedIn);
```
no deocrator, less code, more compatible, and type safe
or potentially:
```TS
isLoggedIn$ = new StoreObservable(AuthState.isLoggedIn);
```
What do you think?

**Comment by dmrickey** [comment]
Yeah both of those look like good options (though I personally like the first option better, it seems a bit weird to me to be "selecting it into a constructor"). How will this work with memoized selectors (or are memoized selectors outside of 3.6.2 already typesafe)?

**Comment by dmrickey** [comment]
Are typesafe selectors in 3.7.x now or is that not back in yet?

**Comment by markwhitfeld** [comment]
@dmrickey Unfortunately not, but we have some really great type safety coming to `createSelector` in v3.8, some really great state slicing utilities (#1653) and potentially some other `@Selector` typing improvements (#1765). 

So, no improvements on the `@Select` decorator as yet, but would you like to try out the experimental utility I made?
ie.: `isLoggedIn$ = new StoreObservable(AuthState.isLoggedIn);`
I would love to get some feedback.

**Comment by dmrickey** [comment]
I can take a look (not sure where the experimental repo is, a quick google didn't turn it up). Does it only work with memoized selectors?

**Comment by markwhitfeld** [comment]
@dmrickey Have a look at this stackblitz:
https://stackblitz.com/edit/ngxs-select-dispatch-utils-v0?file=src%2Fapp%2Fapp.component.ts

What do you think of the `select` and `dispatch` utilities?
Note: these have not yet been tested in production, but I don't see any reason why they would fail.

**Comment by markwhitfeld** [comment]
@dmrickey Ping. I would love to hear what you think.

**Comment by dmrickey** [comment]
> @dmrickey Ping. I would love to hear what you think.

Sorry I never got back to you. I've taken a look and my only question is do all selects specifically need a memoized selector? Can we not select from the state with a syntax similar to the 3.6 (`@Select((x: FooState) => x.foo.bar): Observable<Bar>;`)?

**Comment by dmrickey** [comment]
What's the status of this with the current release?

**Pull request merged**